### PR TITLE
Allow for trusted (connected) organizations

### PR DESCRIPTION
**Pull request by Schparky**
Implemented with a new `organization_trusts` table with two `organizations` foreign key columns. When an `OrganizationTrust` is created, two records are created to make queries slightly more simple and provide future option for asymmetric relationships.

### CODE REVIEW

File: application/models/trust.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+func (t *Trust) ValidateUpdate(tx *pop.Connection) (*validate.Errors, error) {
+	return validate.NewErrors(), nil
+}
+
+// Create stores the Trust data as a new record in the database. Also creates a mirrored copy of the given Trust.
+func (t *Trust) Create() error {
+	var t2 Trust
+	if err := t2.FindByOrgIDs(t.PrimaryID, t.SecondaryID); err == nil {
+		// already exists
+		return nil
```
Comment (Baggerone) [inline application/models/trust.go]: In this case, I wonder if it should still ensure that the mirror exists.

File: application/models/trust.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		return err
+	}
+	return nil
+}
+
+// Remove destroys two Trust records identified by the given Organization IDs
+func (t *Trust) Remove(orgID1, orgID2 int) error {
+	var t1, t2 Trust
+	if err := t1.FindByOrgIDs(orgID1, orgID2); err != nil {
+		return fmt.Errorf("remove Trust failed locating the first Trust record, %s", err)
```
Comment (Baggerone) [inline application/models/trust.go]: In this case, wouldn't we still want to try to remove the mirror?

File: application/gqlgen/mutations.go [hunk 3]
```diff
@@ -162,3 +162,51 @@ func (r *mutationResolver) SetThreadLastViewedAt(ctx context.Context, input SetT
...
+	extras := map[string]interface{}{
+		"user": cUser.UUID,
+	}
+
+	var organization models.Organization
+	if err := organization.FindByUUID(input.PrimaryID); err != nil {
+		return nil, reportError(ctx, err, "CreateTrust.FindPrimaryOrganization", extras)
+	}
+
+	if !cUser.CanEditOrganization(organization.ID) {
```
Comment (Baggerone) [inline application/gqlgen/mutations.go]: Maybe use a new CanCreateTrust function that is more strict.

File: application/locales/global.en.yaml [hunk 4]
```diff
@@ -244,3 +244,17 @@
...
 - id: RemoveWatch.FindByUser
   translation: We had a problem getting a list of remaining Watches
+
+# Trust
+- id: GetOrganizationTrustedOrganizations
+  translation: We had a problem getting a list of trusted organizations
+- id: CreateTrust.FindPrimaryOrganization
+  translation: We had a problem finding the organization to create the Trust
+- id: CreateTrust.Unauthorized
+  translation: We had a problem finding the organization to create the Trust
```
Comment (Baggerone) [inline application/locales/global.en.yaml]: Missing just plain `CreateTrust`?

File: application/actions/organization_fixtures_test.go [hunk 5]
```diff
@@ -238,14 +239,57 @@ func fixturesForOrganizationDomain(as *ActionSuite) OrganizationFixtures {
...
+	orgs[1].Name = "trusted org 1"
+	orgs[2].Name = "trusted org 2"
+	for i := range orgs {
+		orgs[i].AuthType = models.AuthTypeSaml
+		orgs[i].AuthConfig = "{}"
+		orgs[i].Url = nulls.NewString("https://www.example.com")
+		test.MustCreate(as.DB, &orgs[i])
+	}
+
+	trust := models.Trust{PrimaryID: orgs[0].ID, SecondaryID: orgs[1].ID}
```
Comment (Baggerone) [inline application/actions/organization_fixtures_test.go]: The mirror trust should get created too, right? (If so, I guess your desired test results may change.)

File: application/actions/organization_fixtures_test.go [hunk 6]
```diff
@@ -258,7 +302,42 @@ func fixturesForUpdateOrganization(as *ActionSuite) OrganizationFixtures {
...
+	orgs[1].Name = "trusted org 1"
+	orgs[2].Name = "trusted org 2"
+	for i := range orgs {
+		orgs[i].AuthType = models.AuthTypeSaml
+		orgs[i].AuthConfig = "{}"
+		orgs[i].Url = nulls.NewString("https://www.example.com")
+		test.MustCreate(as.DB, &orgs[i])
+	}
+
+	trusts := models.Trusts{
```
Comment (Baggerone) [inline application/actions/organization_fixtures_test.go]: Create the mirrors also?

File: application/models/trust.go (hunk 1 above)
Comment (Schparky) [reply application/models/trust.go]: Yes, good idea.

File: application/models/trust.go (hunk 2 above)
Comment (Schparky) [reply application/models/trust.go]: Yes.

File: application/gqlgen/mutations.go (hunk 3 above)
Comment (Schparky) [reply application/gqlgen/mutations.go]: Sounds good.

File: application/locales/global.en.yaml (hunk 4 above)
Comment (Schparky) [reply application/locales/global.en.yaml]: Right.

File: application/actions/organization_fixtures_test.go (hunk 5 above)
Comment (Schparky) [reply application/actions/organization_fixtures_test.go]: This creates both.

File: application/actions/organization_fixtures_test.go (hunk 6 above)
Comment (Schparky) [reply application/actions/organization_fixtures_test.go]: It does create both.

File: application/models/organizationtrust.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	if err := o.Create(); err != nil {
+		return err
+	}
+
+	mirror := OrganizationTrust{
+		PrimaryID:   o.SecondaryID,
+		SecondaryID: o.PrimaryID,
+	}
+	if err := mirror.Create(); err != nil {
+		_ = DB.Destroy(o)
```
Comment (Baggerone) [inline application/models/organizationtrust.go]: Maybe return the error from here?

File: application/models/organizationtrust.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+	return nil
+}
+
+// Remove destroys the OrganizationTrust record identified by the given Organization IDs
+func (o *OrganizationTrust) Remove(orgID1, orgID2 int) error {
+	var trust OrganizationTrust
+	if err := trust.FindByOrgIDs(orgID1, orgID2); err != nil {
+		if domain.IsOtherThanNoRows(err) {
+			return err
+		} else {
```
Comment (Baggerone) [inline application/models/organizationtrust.go]: with the `return err` above this `else` isn't needed (just its contents)

File: application/gqlgen/mutations.go [hunk 9]
```diff
@@ -162,3 +162,51 @@ func (r *mutationResolver) SetThreadLastViewedAt(ctx context.Context, input SetT
...
+	extras := map[string]interface{}{
+		"user": cUser.UUID,
+	}
+
+	var organization models.Organization
+	if err := organization.FindByUUID(input.PrimaryID); err != nil {
+		return nil, reportError(ctx, err, "RemoveOrganizationTrust.FindPrimaryOrganization", extras)
+	}
+
+	if !cUser.CanCreateOrganizationTrust() {
```
Comment (Baggerone) [inline application/gqlgen/mutations.go]: I'm thinking simple Org Admins should be able to remove an OrgTrust for their own org. @fillup, what do you think?

Comment (fillup) [reply application/gqlgen/mutations.go]: Sure, seems fair. But I'm okay to leave it this way for now too since we wont need that ability for a while.

**Pull request merged**